File: Python/DesignPattern/CreationalDesignPattern/Abstract_FactoryDesign.py
from abc import ABC, abstractmethod


# Product interfaces
class IButton(ABC):
    @abstractmethod
    def render_button(self): pass

class IModal(ABC):
    @abstractmethod
    def render_modal(self): pass

class IScreen(ABC):
    @abstractmethod
    def render_screen(self): pass


# Windows products
class WindowButton(IButton):
    def render_button(self):
        print("Window button rendered")

class WindowModal(IModal):
    def render_modal(self):
        print("Window modal rendered")

class WindowScreen(IScreen):
    def render_screen(self):
        print("Window screen rendered")


# Mac products
class MacButton(IButton):
    def render_button(self):
        print("Mac button rendered")

class MacModal(IModal):
    def render_modal(self):
        print("Mac modal rendered")

class MacScreen(IScreen):
    def render_screen(self):
        print("Mac screen rendered")


# Linux products
class LinuxButton(IButton):
    def render_button(self):
        print("Mac button rendered")

class LinuxModal(IModal):
    def render_modal(self):
        print("Mac modal rendered")

class LinuxScreen(IScreen):
    def render_screen(self):
        print("Mac screen rendered")


# Abstract factory
class IUIFactory(ABC):
    @abstractmethod
    def get_button(self) -> IButton: pass

    @abstractmethod
    def get_modal(self) -> IModal: pass

    @abstractmethod
    def get_screen(self) -> IScreen: pass


class LinuxFactory(IUIFactory):
    def get_button(self) -> IButton:
        return LinuxButton()

    def get_modal(self) -> IModal:
        raise NotImplementedError("Unimplemented method 'get_modal'")

    def get_screen(self) -> IScreen:
        raise NotImplementedError("Unimplemented method 'get_screen'")


class WindowUIFactory(IUIFactory):
    def get_button(self) -> IButton:
        return WindowButton()

    def get_modal(self) -> IModal:
        return WindowModal()

    def get_screen(self) -> IScreen:
        return WindowScreen()


class MacUIFactory(IUIFactory):
    def get_button(self) -> IButton:
        return MacButton()

    def get_modal(self) -> IModal:
        return MacModal()

    def get_screen(self) -> IScreen:
        return MacScreen()


class UIRender:
    def __init__(self, factory: IUIFactory):
        self.button = factory.get_button()
        self.modal = factory.get_modal()
        self.screen = factory.get_screen()
        self.render_ui()

    def render_ui(self):
        self.button.render_button()
        self.modal.render_modal()
        self.screen.render_screen()

    def toggle_ui(self, factory: IUIFactory):
        self.button = factory.get_button()
        self.modal = factory.get_modal()
        self.screen = factory.get_screen()
        self.render_ui()


if __name__ == "__main__":
    my_view = UIRender(WindowUIFactory())
    my_view.toggle_ui(MacUIFactory())

File: C++/DesignPattern/CreationalDesignPattern/Abstract_FactoryDesign.cpp
#include <iostream>
#include <memory>
using namespace std;

// Product interfaces
struct IButton  { virtual void renderButton() = 0; virtual ~IButton() = default; };
struct IModal   { virtual void renderModal()  = 0; virtual ~IModal()  = default; };
struct IScreen  { virtual void renderScreen() = 0; virtual ~IScreen() = default; };

// Windows products
struct WindowButton : IButton { void renderButton() override { cout << "Window button rendered\n"; } };
struct WindowModal  : IModal  { void renderModal()  override { cout << "Window modal rendered\n";  } };
struct WindowScreen : IScreen { void renderScreen() override { cout << "Window screen rendered\n"; } };

// Mac products
struct MacButton : IButton { void renderButton() override { cout << "Mac button rendered\n"; } };
struct MacModal  : IModal  { void renderModal()  override { cout << "Mac modal rendered\n";  } };
struct MacScreen : IScreen { void renderScreen() override { cout << "Mac screen rendered\n"; } };

// Linux products
struct LinuxButton : IButton { void renderButton() override { cout << "Mac button rendered\n"; } };
struct LinuxModal  : IModal  { void renderModal()  override { cout << "Mac modal rendered\n";  } };
struct LinuxScreen : IScreen { void renderScreen() override { cout << "Mac screen rendered\n"; } };

// Abstract factory
struct IUIFactory {
    virtual unique_ptr<IButton> getButton() = 0;
    virtual unique_ptr<IModal>  getModal()  = 0;
    virtual unique_ptr<IScreen> getScreen() = 0;
    virtual ~IUIFactory() = default;
};

struct LinuxFactory : IUIFactory {
    unique_ptr<IButton> getButton() override { return make_unique<LinuxButton>(); }
    unique_ptr<IModal>  getModal()  override { throw runtime_error("Unimplemented method 'getModal'"); }
    unique_ptr<IScreen> getScreen() override { throw runtime_error("Unimplemented method 'getScreen'"); }
};

struct WindowUIFactory : IUIFactory {
    unique_ptr<IButton> getButton() override { return make_unique<WindowButton>(); }
    unique_ptr<IModal>  getModal()  override { return make_unique<WindowModal>(); }
    unique_ptr<IScreen> getScreen() override { return make_unique<WindowScreen>(); }
};

struct MacUIFactory : IUIFactory {
    unique_ptr<IButton> getButton() override { return make_unique<MacButton>(); }
    unique_ptr<IModal>  getModal()  override { return make_unique<MacModal>(); }
    unique_ptr<IScreen> getScreen() override { return make_unique<MacScreen>(); }
};

class UIRender {
    unique_ptr<IButton> button;
    unique_ptr<IModal>  modal;
    unique_ptr<IScreen> screen;
public:
    UIRender(IUIFactory& factory) { toggleUI(factory); }

    void renderUI() {
        button->renderButton();
        modal->renderModal();
        screen->renderScreen();
    }

    void toggleUI(IUIFactory& factory) {
        button = factory.getButton();
        modal  = factory.getModal();
        screen = factory.getScreen();
        renderUI();
    }
};

int main() {
    WindowUIFactory windowFactory;
    MacUIFactory    macFactory;
    UIRender myView(windowFactory);
    myView.toggleUI(macFactory);
    return 0;
}
